### PR TITLE
Add @loopback/security

### PR DESCRIPTION
**Pull request by jannyHou**
Implements https://github.com/strongloop/loopback-next/issues/2900
cherry-picked from https://github.com/strongloop/loopback-next/pull/2902

Create a shared layer for `@loopback/authentication` and `@loopback/authorization`, which defines the security information(context) and binding keys for `Subject` and its members.

For reviewers:

This PR is migrated from draft https://github.com/strongloop/loopback-next/pull/2902.
It contains a comprehensive design of the security system (now`@loopback/authentication` and `@loopback/authorization` only share interface `UserProfile`)

So we have different plans for implementing the security module:

- Plan 1: 
  - If we agree on the design, let's improve and land this PR, then release it
- Plan 2:
  - If it takes time to discuss and agree on the comprehensive design, I can only include interface `Principle` and `UserProfile`(which extends Principle) in this PR.
  - Then incrementally add other pieces in other PRs
- Plan 3:
  - If we couldn't agree on interface `Principle`, let's at least extract the existing `UserProfile` interface here as the first release.

## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-next) 👈

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [ ] API Documentation in code was updated
- [ ] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

### CODE REVIEW

**Comment by dhmlau** [comment]
@jannyHou, for the decreased in code coverage, is there something we can do about it? Thanks.

**Comment by raymondfeng** [comment]
See https://coveralls.io/builds/25161357/source?filename=packages%2Fsecurity%2Fsrc%2Ftypes.ts#L30

File: packages/security/src/__tests__/unit/subject.unit.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+import {DefaultSubject, securityId, UserProfile} from '../..';
+
+describe('DefaultSubject', () => {
+  it('adds users', () => {
+    const subject = new DefaultSubject();
+    const user: UserProfile = {[securityId]: 'user-001', type: 'USER'};
+    subject.addUser(user);
+    expect(subject.user).to.eql(user);
+  });
+});
```
Comment (raymondfeng) [inline packages/security/src/__tests__/unit/subject.unit.ts]: Let's add tests for `addApplication` and `addCredential` etc.

**Comment by jannyHou** [comment]
@dhmlau @raymondfeng Tests added :)

File: packages/security/docs.json [hunk 2]
```diff
@@ -0,0 +1,7 @@
+{
+  "content": [
+    "index.ts",
+    "src/**/*.ts"
+  ],
+  "codeSectionDepth": 4
+}
```
Comment (raymondfeng) [inline packages/security/docs.json]: This file should be removed as we no longer use `strong-docs`.

File: packages/security/package.json [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  "description": "A LoopBack component for security support.",
+  "engines": {
+    "node": ">=8.9"
+  },
+  "scripts": {
+    "acceptance": "lb-mocha \"dist/__tests__/acceptance/**/*.js\"",
+    "build": "lb-tsc",
+    "clean": "lb-clean loopback-authorization*.tgz dist package *.tsbuildinfo",
+    "integration": "lb-mocha \"dist/__tests__/integration/**/*.js\"",
+    "prepublishOnly": "npm run build && npm run build:apidocs",
```
Comment (raymondfeng) [inline packages/security/package.json]: This line should be removed.

File: packages/security/package.json [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  "name": "@loopback/security",
+  "version": "0.0.1",
+  "description": "A LoopBack component for security support.",
+  "engines": {
+    "node": ">=8.9"
+  },
+  "scripts": {
+    "acceptance": "lb-mocha \"dist/__tests__/acceptance/**/*.js\"",
+    "build": "lb-tsc",
+    "clean": "lb-clean loopback-authorization*.tgz dist package *.tsbuildinfo",
```
Comment (raymondfeng) [inline packages/security/package.json]: It should be:
```
"clean": "lb-clean loopback-security*.tgz dist package *.tsbuildinfo",
```

File: packages/security/package.json [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+  "dependencies": {
+    "@loopback/context": "^1.13.0",
+    "@loopback/core": "^1.8.5",
+    "debug": "^4.1.1"
+  },
+  "devDependencies": {
+    "@loopback/build": "^1.5.2",
+    "@loopback/testlab": "^1.2.7",
+    "@types/debug": "^4.1.4",
+    "@types/node": "10.14.15",
```
Comment (raymondfeng) [inline packages/security/package.json]: Should it be a range?

File: packages/security/package.json [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    "README.md",
+    "index.js",
+    "index.d.ts",
+    "dist",
+    "src",
+    "!*/__tests__"
+  ],
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/strongloop/loopback-next.git"
```
Comment (raymondfeng) [inline packages/security/package.json]: Add `"directory": "packages/security"`

File: packages/security/package.json [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+  }
+}
```
Comment (raymondfeng) [inline packages/security/package.json]: Let's compare with an existing module such as `@loopback/context` to make sure the package.json is consistent.

File: packages/security/package.json (hunk 5 above)
Comment (jannyHou) [reply packages/security/package.json]: good catch 👍

**Pull request merged**